flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -1320,8 +1339,46 @@ class ScalarExprLowering {
 
     for (const auto &arg : caller.getPassedArguments()) {
       const auto *actual = arg.entity;
-      if (!actual)
-        TODO("optional argument lowering");
+      if (!actual) {
+        // absent optional.
+        auto argTy = caller.genFunctionType().getInput(arg.firArgument);
+        if (arg.passBy == PassBy::Box) {
+          // TODO: for now create a scalar box with null address to implement
+          // absent optional (it will be casted to the correct box type later,
+          // do not bother making non trivial embox for that).
+          // This will waste a bit of time/memory compare to
+          // using a null descriptor address, but the fact that box are passed
+          // as ssa values in fir makes it a bit harder to implement absent
+          // non-pointer/non-allocatable descriptor as a null descriptor address
+          // (the issue is mainly on the caller side to check if a fir.box is
+          // null).
+          // In the bind(c) context, it will be required to handle optional
+          // descriptor as a null descriptor address (F2018 18.3.6 point 7).
+          auto null = builder.createNullConstant(
+              loc, builder.getRefType(builder.getI32Type()));
```

Comment:
I thought `fir.zero !fir.box<T>` would be used to create a 0 descriptor.
Also, can't a value of box type be casted to a ref?

Reply:
> I thought fir.zero !fir.box<T> would be used to create a 0 descriptor.

It seems `fir.zero !fir.box<T>` will create a null descriptor address as opposed to a descriptor with all data being set to zero (looking at the codegen [here](https://github.com/flang-compiler/f18-llvm-project/blob/21c6631db427843fb89171958f6978a45201732c/flang/lib/Optimizer/CodeGen/CodeGen.cpp#L2279-L2281)), which is not exactly what I want right now.

> Also, can't a value of box type be casted to a ref?

It's not allowed [in the fir.convert verifier](https://github.com/flang-compiler/f18-llvm-project/blob/21c6631db427843fb89171958f6978a45201732c/flang/include/flang/Optimizer/Dialect/FIROps.td#L3129-L3137). Just removing the check would make it work, and make my life easier here, but is it ok to expose in FIR that `fir.box` are pointers under the hood ?